tests/test_charms.py

```diff
@@ -0,0 +1,251 @@
+from pathlib import Path
+from random import choices
+from string import ascii_lowercase
+from subprocess import check_output, run
+from time import sleep
+from urllib.parse import urlparse
+
+import pytest
+import yaml
+from selenium.common.exceptions import JavascriptException, WebDriverException
+from selenium.webdriver.common.by import By
+from selenium.webdriver.firefox.options import Options
+from selenium.webdriver.support import expected_conditions as EC
+from selenium.webdriver.support.ui import WebDriverWait
+from seleniumwire import webdriver
+
+CONTROLLER_METADATA = yaml.safe_load(Path("charms/jupyter-controller/metadata.yaml").read_text())
+UI_METADATA = yaml.safe_load(Path("charms/jupyter-ui/metadata.yaml").read_text())
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test):
+    run(["juju", "switch", ops_test.model_full_name])
+    await ops_test.model.deploy("cs:istio")
+    await ops_test.model.applications["istio-pilot"].set_config(
+        {"default-gateway": "kubeflow-gateway"}
+    )
+    sleep(10)
+    run(
```

Comment:
This won't be necessary shortly, after https://github.com/canonical/istio-operators/pull/31 is merged and pushed to the charm store. More of an FYI, since this PR will probably get merged before that all happens.